bucketeer/src/main/kotlin/io/bucketeer/sdk/android/internal/event/db/EventDaoImpl.kt

```diff
@@ -26,9 +26,22 @@ internal class EventDaoImpl(
   }
 
   override fun addEvents(events: List<Event>) {
+    // This approach below is a reference from iOS SDK.
+    // It could be better but This approach will create a minimum impact,
+    // And because the number of pending events in the database is small
+    // So I think we are safe to do this without changing too much
+    // 1. Get all current events and collect hash
+    // https://kotlinlang.org/docs/data-classes.html
+    val storedEvents = getEvents()
+    val storedEventHashList = storedEvents.map {
+      it.event.hashCode()
+    }
     sqLiteOpenHelper.writableDatabase.transaction {
-      events.forEach { event ->
-        addEventInternal(this, event)
+      events.forEach { item ->
+        // 2. Push to the database when the event data do not exist in the database
+        if (!storedEventHashList.contains(item.event.hashCode())) {
```

Comment:
Thank @yshrsmz , let me test this case

Reply:
@cre8ivejp Could you elaborate on the requirements?

How should we dedupe the events? 

- Should we dedupe all event types?
- Should we keep the first saved event or overwrite it with the latest one?
- Should we dedupe if all fields are equal, or just check ApiId & event type?

Reply:
@yshrsmz  my current implementation will dedupe all events (metrics, goal, evaluation) like a reference implement from iOS SDK.
as the requirement is for `metrics` only, I should add a filter logic.
But I will wait for a final confirmation from cre8ivejp

Reply:
@cre8ivejp Also, Also, I see one case the `metric event` in the `cache database` may not be equal, but they seem duplicate.
**Because They will have a different timestamp while another field could be equal.**

For example 
```kotlin
Event(
    id = "a85e88817fec",
    type = EventType.METRICS,
    event = EventData.LatencyMetricEvent(
      timestamp = 1661780821,
      userId = user1.id,
      latency=2000,
      user = user1,
      value = 0.0,
      tag = "",
      sourceId = SourceID.ANDROID,
      sdkVersion = io.bucketeer.sdk.android.BuildConfig.SDK_VERSION,
      metadata = mapOf(
        "app_version" to "1.2.3",
        "os_version" to "os_version_value",
        "device_model" to "device_model_value",
      ),
    ),
  )
  ```
  
  ```kotlin
  Event(
    id = "408741bd",
    type = EventType.METRICS,
    event = EventData.LatencyMetricEvent(
      timestamp = 1661780900,
      userId = user1.id,
      latency=2000,
      user = user1,
      value = 0.0,
      tag = "",
      sourceId = SourceID.ANDROID,
      sdkVersion = io.bucketeer.sdk.android.BuildConfig.SDK_VERSION,
      metadata = mapOf(
        "app_version" to "1.2.3",
        "os_version" to "os_version_value",
        "device_model" to "device_model_value",
      ),
    ),
  )
  ```

Should we exclude the `timestamp` when preventing the duplicate?

Reply:
@yshrsmz @duyhungtnn 

> Should we dedupe all event types?

No. Only metrics events. Otherwise, it could affect other services on the server side.

> Should we keep the first saved event or overwrite it with the latest one?

Keep the first event. Is this the same condition in JS SDK?

> Should we dedupe if all fields are equal, or just check ApiId & event type?

ApiId and event type is enough for me. What is the JS SDK's condition for this situation?

Reply:
Hi @yshrsmz  @cre8ivejp  I added logic to filter only metric events and update the test case.
Please help me to take a look https://github.com/bucketeer-io/android-client-sdk/pull/68/commits/113fea0693db40811abb450961b3869df516bd2b

Reply:
@cre8ivejp Thanks, but you don't need to check the whole EventData. ApiId & protobufType is enough.

For example, when you save a new `type.googleapis.com/bucketeer.event.client.LatencyMetricsEvent` for `ApiId.GET_EVALUATIONS`, we search for an existing event that matches these 2 conditions. We don't need to save a new event if the event already exists in the cache.

I was thinking about adding `apiId` and `eventDataType` field to `event` table so that we can check existing events with SQL query.

Reply:
@duyhungtnn  Thanks, but you don't need to check the whole EventData. ApiId & protobufType is enough.

For example, when you save a new `type.googleapis.com/bucketeer.event.client.LatencyMetricsEvent` for `ApiId.GET_EVALUATIONS`, we search for an existing event that matches these 2 conditions. We don't need to save a new event if the event already exists in the cache.

I was thinking about adding `apiId` and `eventDataType` field to `event` table so that we can check existing events with SQL query.

Reply:
I think `apiId` and `eventDataType` are not enough in the case of `LatencyMetric` (latencySecond)  and `SizeMetric` (size_byte)
@yshrsmz 
cc: @cre8ivejp Please check this case, and let me know the final requirement for apply

Reply:
> Regarding JS SDK implementation, JS SDK does not dedupe events for now, as its spec was not clear when I developed the SDK

@yshrsmz, thanks for letting me know. I will create an issue in the JS SDK repository.

> I think apiId and eventDataType are not enough in the case of LatencyMetric (latencySecond) and SizeMetric (size_byte)

@duyhungtnn, while checking only the `EventType.METRICS` would work, checking the ApiId and the protobufType would make more sense because we don't need to check the whole object. So for this case, ApiId and protobufType would be enough even for LatencyMetric and SizeMetric events. I didn't understand what scenario it wouldn't.

The interval for `get_evaluations` requests is 10 minutes. Usually, those events won't be in the DB too long because the `register_events` will try to report them every minute. Even if the internet is offline, it won't create new events for `get_evaluations`.

@yshrsmz @duyhungtnn 

I said before to keep the first event but to be honest, it really doesn't matter as long the server gets an event.
So perhaps we could just implement it to upsert the event in the DB when it is a metrics event. What do you guys think?

Reply:
The main reason we want to avoid saving the same metric events is that when the user uses the application in offline mode, it will keep creating error metrics events unnecessarily. We could also check the network status and stop making requests to the server, but for Android, we would need the `ACCESS_NETWORK_STATE` permission, and as an SDK, this isn't a good practice.

For iOS, we wouldn't need permission to check the network status, but it would create inconsistency between Android and iOS. If you guys can come up with a better idea to solve this, I would appreciate it.

Reply:
thank @cre8ivejp  for letting me know more context for more understanding 

Without `ACCESS_NETWORK_STATE` permission, I think we could check at an interval if 
1. we could make success simple request as a "ping request" to our server 
2. DNS lookup to server domain is working.

But I believe the current approach in this PR may more simple than, and make sure no duplicates.

Reply:
I will be update this PR with `main` branch for access new property `ApiId`  and apply the changes

Reply:
I updated this PR, please help me to continue reviewing.
@yshrsmz  @cre8ivejp